pycqed/measurement/detector_functions.py

```diff
@@ -192,23 +192,84 @@ def finish(self):
 
 
 class IndexDetector(Detector_Function):
+    """Detector function that indexes the result of another detector function.
+
+    Args:
+        detector:
+            detector function that returns multiple values per sweep point
+        index:
+            Index if the element returned from the original detector function
+            output. Can be an integer or a tuple. If an integer, then value is
+            interpreted as a channel index of the original detector function.
+            In case of a tuple, the first value corresponds to a channel index
+            and the second value corresponds to an index in the original
+            hardware sweep. Using a tuple converts a hardware sweep to a
+            software sweep.
+    """
+
     def __init__(self, detector, index):
         super().__init__()
         self.detector = detector
         self.index = index
         self.name = detector.name + '[{}]'.format(index)
-        self.value_names = [detector.value_names[index]]
-        self.value_units = [detector.value_units[index]]
+        if isinstance(self.index, tuple):
+            self.value_names = [detector.value_names[index[0]]]
+            self.value_units = [detector.value_units[index[0]]]
+            self.detector_control = 'soft'
+        else:
+            self.value_names = [detector.value_names[index]]
+            self.value_units = [detector.value_units[index]]
+            self.detector_control = detector.detector_control
+            self.index = [index]
+
+    def prepare(self, **kw):
+        self.detector.prepare(**kw)
+
+    def get_values(self):
+        v = self.detector.get_values()
+        for i in self.index:
+            v = v[i]
+        return v
```

Comment:
Thanks for the clean-up. Looking at the refactored code, I am still trying to make sense of this loop, and I have a follow-up question. Isn't this just
```
return self.detector.get_values()[self.index[-1]]
```
? If yes, I would find this explicit version more readable (and it should be negligibly more efficient). If no, we should add a comment to explain the loop to people like me ;-). Same in acquire_data_point.